src/pybind/mgr/ansible_orchestrator/module.py

```diff
@@ -0,0 +1,364 @@
+
+        The result of the playbook execution can be customized through the
+        function provided as 'process_output' attribute
+
+        @return string: Result of the operation formatted if it is possible
+        """
+
+        processed_result = []
+
+        if self._is_complete:
+            raw_result = self.pb_execution.get_result(self.event_filter)
+
+            if self.process_output:
+                processed_result = self.process_output(
+                                            raw_result,
+                                            self.ar_client,
+                                            self.pb_execution.play_uuid)
+            else:
+                processed_result = raw_result
+
+        self._result = processed_result
+
+
+class AnsibleChangeOperation(orchestrator.WriteCompletion):
+    """Operations that changes the "cluster" state
+
+    Modifications/Changes (writes) are a two-phase thing, firstly execute
+    the playbook that is going to change elements in the Ceph Cluster.
+    When the playbook finishes execution (independently of the result),
+    the modification/change operation has finished.
+    """
+    def __init__(self):
+        super(AnsibleChangeOperation, self).__init__()
+
+        self.error = False
+    @property
+    def status(self):
+        """Return the status code of the operation
+        """
+        #TODO
+        return 0
+
+    @property
+    def is_persistent(self):
+        """
+        Has the operation updated the orchestrator's configuration
+        persistently?  Typically this would indicate that an update
+        had been written to a manifest, but that the update
+        had not necessarily been pushed out to the cluster.
+
+        In the case of Ansible is always False.
+        because a initiated playbook execution will need always to be
+        relaunched if it fails.
+        """
+
+        return False
+
+    @property
+    def is_effective(self):
+        """Has the operation taken effect on the cluster?
+        For example, if we were adding a service, has it come up and appeared
+        in Ceph's cluster maps?
+
+        In the case of Ansible, this will be True if the playbooks has been
+        executed succesfully.
+
+        @return Boolean: if the playbook has been executed succesfully
+        """
+
+        return self.status == ExecutionStatusCode.SUCCESS
+
+    @property
+    def is_errored(self):
+        return self.error
+
+    @property
+    def is_complete(self):
+        return self.is_errored or (self.is_persistent and self.is_effective)
+
+
+class Module(MgrModule, orchestrator.Orchestrator):
+    """An Orchestrator that an external Ansible runner service to perform
+    operations
+    """
+
+    OPTIONS = [
+        {'name': 'server_addr'},
+        {'name': 'server_port'},
+        {'name': 'username'},
+        {'name': 'password'},
+        {'name': 'certificate'}  # Ansible runner https server certificate file
```

Comment:
Is this a filename?  Manager modules should not depend on files out on disk, rather the certificate should be stored like the username/password (see how this is done for server side certs in dashboard)

Reply:
Ok. I will check the dashboard implementation and i will follow your advice.
But this is not part of this PR.. so i prefer to implement it later. I wouldn't  like to add features over features in an endless PR